Add `rstuf artifact repository` command

ivanayov opened the pull request:
This change adds an artifact repostority command that allows updating artifacts repository configuration used for downloads


<!--- Please reference the issue(s) this PR fixes. -->
Fixes #332 
Fixes #434 

# Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

# Additional requirements
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature


# Code of Conduct
By submitting this PR, you agree to follow our [Code of Conduct](CODE_OF_CONDUCT.rst).

- [x] I agree to follow this project's Code of Conduct

ivanayov commented:
Codecov complains about dynaconf method not being tested. We should rather ignore that, as we don't want and don't need test coverage for an external library.

ivanayov commented:
@MVrachev @kairoaraujo @KAUTH Could some of you please review? Thanks :)

MVrachev commented:
@ivanayov when you fix an issue please add a new commit, do not amend to it.
That way we can see the changes.
In the end, we will squash everything, but in RSTUF projects we have the practice of adding new commits instead of rebasing.

kairoaraujo commented on docs/source/guide/index.rst:
```diff
@@ -914,6 +914,90 @@ This command adds the provided artifact to the TUF Metadata using the RSTUF REST
+
+    Repository management.
+
+    ╭─ Options ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
+    │ --help  -h    Show this message and exit.                                                                                                                                 │
+    ╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
+    ╭─ Commands ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
+    │ delete                           Delete repository.                                                                                                                       │
+    │ set                              Set a new repository.                                                                                                                    │
+    │ show                             List configured repositories.                                                                                                            │
+    │ update                           Update repository.                                                                                                                       │
+    │ use                              Switch current repository.                                                                                                               │
+    ╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```
Nit: Copy and paste it from a smaller (length) terminal. It will have less space on the left.


I would rename the commands here
`set` -> `add` 
It will be aligned with other subcomands, and it sounds more specific.

`use` -> `set`
to set the current repository

kairoaraujo commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,261 @@
+@repository.command()
+@click.pass_context
+@click.argument("repository")
+def use(context: Context, repository) -> None:
```
```suggestion
def use(context: Context, repository: str) -> None:
```

kairoaraujo commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,261 @@
+    metadata_url: str,
+    artifacts_url: str,
+    hash_prefix: bool,
+    repository,
```
Nit: Isn't it better for the `repository` to be an explicit parameter `--name` in the `set` (I would prefer `add` as I mentioned above)

```suggestion
    repository: str,
```

ivanayov replied:
We previously agreed on `rstuf artifact repository add <repo_name>`, but if you prefer a flag, I don't mind changing the layout.

kairoaraujo replied:
I still think that having `add [REPOSITORY]` is confusing for the user
![screenshot_2024-01-25_at_18 06 54_720](https://github.com/repository-service-tuf/repository-service-tuf-cli/assets/546882/8e4cc21e-010d-4874-8e2d-76a247e063c9)
t leads the user to add an empty repository of information :confused:
Another thing I noticed is that I don't have a way to check my settings without opening the file
![screenshot_2024-01-25_at_18 07 37_480](https://github.com/repository-service-tuf/repository-service-tuf-cli/assets/546882/0655f5c0-61af-48d4-8b5c-43eb54568b24)


Another thing is that the `rstuf admin artifacts repository` should list all existent repository names and the default one should be highlighted.

 Otherwise, I need to remember my repositories' names.

kairoaraujo commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,261 @@
+def show(
+    context: Context,
+    all: bool,
+    repository,
```
repository: string
```suggestion
    repository: str,
```

ivanayov replied:
Same here - if we agree to change the initial layout, I'll update it.

kairoaraujo commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,261 @@
+
+    rstuf_config = context.obj.get("settings")
+
+    if all:
+        if rstuf_config.get("REPOSITORIES"):
+            for repo in rstuf_config.get("REPOSITORIES"):
+                console.print(repo)
+        else:
+            console.print("There are no configured repositories")
+            return
+
+    if repository:
+        if rstuf_config.get("REPOSITORIES") and rstuf_config[
+            "REPOSITORIES"
+        ].get(repository):
+            console.print(rstuf_config["REPOSITORIES"].get(repository))
+        else:
+            console.print(
+                f"Repository {repository} is missing in your configuration"
+            )
```
I think if the repository is empty we can also show the current repository details.

```suggestion
    if all:
        if rstuf_config.get("REPOSITORIES"):
            for repo in rstuf_config.get("REPOSITORIES"):
                console.print(repo)
        else:
            raise click.ClickException("There are no configured repositories")

    elif repository:
        if rstuf_config.get("REPOSITORIES") and rstuf_config[
            "REPOSITORIES"
        ].get(repository):
            console.print(rstuf_config["REPOSITORIES"].get(repository))
        else:
            raise click.ClickException(
                f"Repository {repository} is missing in your configuration"
            )
    else:
        console.print(
            rstuf_config["REPOSITORIES"].get(
                rstuf_config["CURRENT_REPOSITORY"])
        )
            
```

kairoaraujo commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,261 @@
+    rstuf_config = context.obj.get("settings").as_dict()
+
+    if not rstuf_config.get("REPOSITORIES"):
+        console.print("There are no configured repositories to update")
+        return
```
```suggestion
        raise click.ClickException("There are no configured repositories to update")
```

kairoaraujo commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,261 @@
+        return
+
+    if not rstuf_config["REPOSITORIES"].get(repository):
+        console.print(
+            f"Repository {repository} not available in config. "
+            "You can create it instead"
+        )
+        return
```
```suggestion
        raise click.ClickException(
            f"Repository {repository} not available in config. "
            "You can `add` it instead"
        )
```

kairoaraujo commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,261 @@
+@repository.command()
+@click.pass_context
+@click.argument("repository")
+def delete(context: Context, repository) -> None:
```
```suggestion
def delete(context: Context, repository: str) -> None:
```

kairoaraujo commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,261 @@
+    metadata_url: str,
+    artifacts_url: str,
+    hash_prefix: bool,
+    repository,
```
```suggestion
    repository: str,
```

kairoaraujo commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,261 @@
+    rstuf_config = context.obj.get("settings").as_dict()
+
+    if not rstuf_config.get("REPOSITORIES"):
+        console.print(
+            "There are no configured repositories. Nothing to delete"
+        )
+        return
```
```suggestion
        raise click.ClickException(
            "There are no configured repositories. Nothing to delete"
        )
```

kairoaraujo commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,261 @@
+        return
+
+    if not rstuf_config["REPOSITORIES"].get(repository):
+        console.print(
+            f"Repository {repository} not available. Nothing to delete"
+        )
+        return
```
```suggestion
        raise click.ClickException(
            f"Repository {repository} not available. Nothing to delete"
        )
```

kairoaraujo reviewed (changes requested):
Hi @ivanayov 
Thank you for your PR. You did a great work here. 🎉

MVrachev commented:
I will also have a look at the pr and will discuss with @ivanayov what you two have decided as the design and flow.

ivanayov commented:
Hi @kairoaraujo , CI is now fixed and comments are up-to-date

ivanayov commented:
push-forced latest updates from main

MVrachev commented on docs/source/guide/index.rst:
```diff
@@ -938,6 +938,86 @@ This command allows downloading an artifact from a provided repository using the
+    │ --help  -h    Show this message and exit.                                                                                                                                 │
+    ╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
+    ╭─ Commands ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
+    │ delete                           Delete repository.                                                                                                                       │
```
For consistency with `Add a new repository`:
```suggestion
    │ delete                           Delete a repository.                                                                                                                       │
```

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+@click.argument("repository")
+def delete(context: Context, repository: str) -> None:
+    """
+    Delete repository.
```
For consistency with `Add a new repository`:
```suggestion
    Delete a repository.
```

MVrachev commented on docs/source/guide/index.rst:
```diff
@@ -938,6 +938,86 @@ This command allows downloading an artifact from a provided repository using the
+    ❯ rstuf artifact repository delete --help
+
+    Usage: rstuf artifact repository delete [OPTIONS] REPOSITORY
+
```
Nit: add a line using the descriptions like `Delete a repository`.
Here is how it looks currently:
<img width="1118" alt="image" src="https://github.com/repository-service-tuf/repository-service-tuf-cli/assets/16246778/95c452a9-e62e-43e2-9b78-abf69d3bef21">

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+from repository_service_tuf.cli.artifact import artifact
+
+
+def write_config(settings_path, settings_data, merge=True):
```
Maybe the default could be `merge=False` as everywhere you call it with `merge=False`. :))

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+
+
+def write_config(settings_path, settings_data, merge=True):
+    write(settings_path, settings_data, merge)  # pragma: no cover
```
What error did you stop with `pragma`?

ivanayov replied:
"no cover", i.e. a test coverage check, as it's an external method we don't want to test, but reduces codecov %.

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+@click.option(
+    "-n",
+    "--name",
+    help="A repository name.",
```
```suggestion
    help="The repository name.",
```

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+@click.option(
+    "-r",
+    "--root",
+    help="A metadata URL to the initial trusted root or a local file.",
```
```suggestion
    help="The metadata URL to the initial trusted root or a local file.",
```

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+@click.option(
+    "-a",
+    "--artifacts-url",
+    help="An artifacts base URL to fetch from.",
```
```suggestion
    help="The artifacts base URL to fetch from.",
```

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+@click.option(
+    "-p",
+    "--hash-prefix",
+    help="A hash prefix.",
```
```suggestion
    help="Whether to add a hash prefix to artifact names.",
```

MVrachev commented on docs/source/guide/index.rst:
```diff
@@ -938,6 +938,86 @@ This command allows downloading an artifact from a provided repository using the
+    Add a new repository.
+
+    ╭─ Options ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
+    │ *  --name           -n  TEXT  A repository name. [required]                                                                                                               │
+    │ *  --root           -r  TEXT  A metadata URL to the initial trusted root or a local file. [required]                                                                      │
+    │ *  --metadata-url   -m  TEXT  TUF Metadata repository URL. [required]                                                                                                     │
+    │ *  --artifacts-url  -a  TEXT  An artifacts base URL to fetch from. [required]                                                                                             │
+    │    --hash-prefix    -p        A hash prefix.                                                                                                                              │
```
Please update the docs based on my comments in the repository module.

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+        console.print(f"Repository {name} already configured")
+        success_msg = f"Successfully updated repository {name}"
+
+    encoded_root = b""
+    if root:
+        encoded_root = base64.b64encode(bytes(root, "utf-8"))
```
Why did you decide to encode the root?

ivanayov replied:
We wouldn't like to keep the trusted root path in plane text in our config

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+@click.option(
+    "-r",
+    "--root",
+    help="A metadata URL to the initial trusted root or a local file.",
```
```suggestion
    help="The metadata URL to the initial trusted root or a local file.",
```

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+@click.option(
+    "-a",
+    "--artifacts-url",
+    help="An artifacts base URL to fetch from.",
```
```suggestion
    help="The artifacts base URL to fetch from.",
```

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+@click.option(
+    "-p",
+    "--hash-prefix",
+    help="A hash prefix.",
```
```suggestion
    help="Whether to add a hash prefix to artifact names.",
```

MVrachev commented on docs/source/guide/index.rst:
```diff
@@ -938,6 +938,86 @@ This command allows downloading an artifact from a provided repository using the
+    Update repository.
+
+    ╭─ Options ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
+    │ --root           -r  TEXT  A metadata URL to the initial trusted root or a local file.                                                                                    │
+    │ --metadata-url   -m  TEXT  TUF Metadata repository URL.                                                                                                                   │
+    │ --artifacts-url  -a  TEXT  An artifacts base URL to fetch from.                                                                                                           │
+    │ --hash-prefix    -p        A hash prefix.                                                                                                                                 │
```
Please update as the changes suggested in the repository module

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+        )
+
+    if root:
+        rstuf_config["REPOSITORIES"][repository][  # pragma: no cover
```
Maybe you don't need `pragma`

ivanayov replied:
I have test coverage for this use-case, but codecov doesn't find it properly.

MVrachev replied:
I couldn't find a test where you test that.
I think you should add a test for that.

MVrachev commented on repository_service_tuf/cli/artifact/repository.py:
```diff
@@ -0,0 +1,278 @@
+    if metadata_url:
+        rstuf_config["REPOSITORIES"][repository]["metadata_url"] = metadata_url
+    if artifacts_url:
+        rstuf_config["REPOSITORIES"][repository][  # pragma: no cover
```
Maybe you don't need `pragma`

MVrachev reviewed (changes requested):
Great job overall!
I left small comments.

Maybe it's a good idea to have an example config file under this folder:
https://github.com/repository-service-tuf/repository-service-tuf-cli/tree/main/tests/files

Also, it's probably good to mention where this `config` file will be located if `--config` is not provided as an argument.

Note: I didn't review the tests yet.

ivanayov commented:
Thanks @MVrachev.  Regarding the example config, I think it fits better in docs, rather than in tests.

ivanayov commented:
Force-pushed due to rebase from main.

As we agreed on - I added only inline docs for the expected config. The user doesn't need to know details about the config file itself. One can see read about the `--config` setup in the already existing docs. For the purpose and scope of the `repository` command I tested operating with an empty file without any pre-configuration. One can easily add repositories, read, update and delete them, as well as set the default one without any manual interaction with the config file.

Added a fix for `trusted_root`, as when it was added with the `repository add` command, it was prefixed with `!!binary |` causing parsing errors later.

MVrachev commented on tests/unit/cli/artifact/test_repository.py:
```diff
@@ -0,0 +1,732 @@
+        assert check_settings.get("CURRENT_REPOSITORY") == "r2"
+        assert test_result.exit_code == 0
+
+    def test_repository_set_one_param_only(
```
I don't see a point of this test specifically as we test click functionality to check if the required arguments are there.

MVrachev reviewed (changes requested):
I reviewed the tests as well.
There are a couple of comments from my last review that were not addressed.
Could you please have a look at them as well?

ivanayov commented:
@MVrachev @kairoaraujo It's all addressed now :) I added a small bug fix as well.

MVrachev reviewed (approved):
A lot of work went on that pr, but it's well documented and tested.

There is only one question that I didn't get an answer for:
In one of my comments I asked when you encode the root in base64 are you sure you can it then be loaded back again to a Metadata object?
Have you tried it?

ivanayov commented:
@MVrachev  We don't load Metadata object in none of `add`/`update` - we're working directly with the config. Your comment refers to `add`, where `encoded_root` is already decoded prior being added to the config dict - just 5 lines bellow the encoding you refer to.

ivanayov commented:
push-forced a rebase from `main`
